# üçÉ Li-Fa-Fang-Yao

**`Li-Fa-Fang-Yao`** is a big semantic netowrk dataset of traditional Chinese medicine (TCM).

> Currently, we have open-sourced all semantic types of nodes and publicly available relationship data entries. Due to the large size of the complete entities and relations data, it cannot be uploaded to GitHub. If you need it, please send an email to mc36401@um.edu.mo to explain your purpose.

## 1.Statistics

| Semantic Type| Resource | Total |
| --- | --- | --- |
| Symptoms  | GB/T16751-1-2023-Disease, SymMap | 2444 |
| Diseases  | GB/T16751-1-2023-Disease, UMLS   | 1167 |
| Syndromes | GB/T16751-2-2021-Syndrome | 1648 |
| Therapys  | GB/T16751-3-2023-Therapeutics | 764 |
| Formulaes | GBT31773-2015-Formulae | 1089 |
| Herbs     | GBT31774-2015-Herbs, HERB, ITCM, TCMIO | 1603 |
| Ingredients | PubChem, HERB  | 1283 |
| References  | PubMed  | 6705 |
| Experiments | HERB  | 2231 |
| Relation | References, Literature  | 27185|

## 2.Running

Please run `buildKG.py` to construct `neo4j graph database`, 

Replace with your username and password in the `neo4j_connector` function.

```python
...
def neo4j_connector(
    uri="bolt://localhost:7687", 
    username="neo4j", # Replace with your username
    password="neo4j@soap" # and password
):
    """
    Connect to Neo4j instance.
    """
    driver = GraphDatabase.driver(uri, auth=(username, password))
    return driver
```

## 3.Details

### Symptoms - 2444

- name_cn: standard Chinese name of the symptom (e.g., "Â§¥Áóõ" [Headache], "ÂèëÁÉ≠" [Fever]).
- name_en: standard English name of the symptom (e.g., "Headache", "Fever").
- name_pinyin: Pinyin spelling of the Chinese symptom name
- clinical_performance: Clinical manifestations or specific characteristics of the symptom (e.g., "paroxysmal stabbing pain", "accompanied by nausea").
- body_part: Anatomical location associated with the symptom (e.g., "head", "abdomen").
- indicator: Relevant clinical biomarkers or diagnostic indicators / Chinese syndrome summary. (e.g., "elevated leukocyte count", "positive CRP").
  
### Diseases - 1167

- name_cn: standard traditional Chinese disease name
- name_en: The translation of the Chinese disease name into English
- pathogenesis: Pathological mechanisms or etiology
- clinical_performance: Typical clinical features
- Disease_name: The official or primary name by which the disease is commonly known.
- Disease_alias_name: A list of alternative names, synonyms, or aliases for the disease, separated by semicolons.
- DisGeNET_disease_type: The type or classification of the disease as categorized in the DisGeNET database.
- UMLS_disease_type: The type of disease as classified in the Unified Medical Language System (UMLS), a comprehensive system for biomedical vocabularies.
- MeSH_disease_class: The class or category of the disease within the Medical Subject Headings (MeSH), which is used for indexing, cataloging, and searching biomedical literature.
- HPO_disease_class: The class or category of the disease within the Human Phenotype Ontology (HPO), which provides a standardized vocabulary for human phenotypic abnormalities.
- DO_disease_class: The class or category of the disease within the Disease Ontology (DO), which offers a classification for human diseases.
- UMLS_disease_type_id: The identifier for the disease type within the UMLS classification system.
- MeSH_disease_class_id: The identifier for the MeSH disease class.
- HPO_disease_class_id: The identifier for the HPO disease class.
- DO_disease_class_id: The identifier for the DO disease class.
- HERB_id: The cross-reference identifier for the disease within the HERB database.
- DisGeNET_id:  The unique identifier for the disease within the DisGeNET database.
- MeSH_id: The cross-reference identifier for the disease within the MeSH database.
- HPO_id: The cross-reference identifier for the disease within the HPO database.
- DO_id: The cross-reference identifier for the disease within the Disease Ontology database.
- ICD10_id: The cross-reference identifier for the disease within the International Classification of Diseases, 10th Revision (ICD-10), which is used for epidemiological purposes and for the clinical coding of diseases.
- OMIM_id: The cross-reference identifier for the disease within the Online Mendelian Inheritance in Man (OMIM) database, which focuses on the genetic aspects of human diseases.

### Syndromes - 1648

- name_cn: standard traditional Chinese syndrome name
- name_en: The translation of the Chinese syndrome name into English
- pathogenesis: Pathological mechanisms or etiology
- clinical_performance: Typical clinical features

### Therapys - 764

- name_cn: Standard Chinese therapy name
- name_en: The translation of the Chinese therapy name into English
- description: Technical details or principles of the therapy
- alias: Alternative names or abbreviations 

### Formulaes - 1089

- name_cn: Standard Chinese formula name 
- source: Historical or literary origin 
- efficacy_class: Pharmacological classification
- composition: Herbal ingredients
- function: Primary therapeutic effects
- indications: Indications for use
- alias: Alternative names

### Herbs - 1603

The herb info file includes all detail information of the 1,603 herbs. The detail of each column are as follows:

- name_cn: The official Chinese name of the herb.
- name_en: The common English name of the herb.
- name_pinyin: The pronunciation of the herb's Chinese name using Pinyin.
- name_latin: The scientific Latin name that classifies the species of the herb.
- name_alias: The alternate names or synonyms of the herb.
- property_cn: A description of the inherent characteristics or 'properties' of the herb, often referring to its nature in TCM, such as being warm, cold, sweet, pungent, etc.
- property_en: The common English translation of the 'property_cn' column.
- meridian_cn: The specific pathways or 'meridians' in the body through which the herb is believed to act, according to TCM theory in Chinese.
- meridian_en: The common English translation of the 'meridian_cn' column.
- usepart_cn: The specific part of the herb that is used for medicinal purposes, such as the root, leaf, seed, or bulb in Chinese.
- usepart_en: The common English translation of the 'usepart_cn' column.
- function_cn: A description of the medicinal functions or 'effects' the herb is traditionally believed to have.
- function_en: The common English translation of the 'function_cn' column.
- indication_cn: The conditions or symptoms for which the herb is traditionally used to treat.
- indication_en: The common English translation of the 'indication_cn' column.
- clinical_manifestations: The observed clinical effects or indications that the herb may address or alleviate.
- toxicity_cn: Information regarding the potential toxicity or side effects of the herb in Chinese.
- toxicity_en: The common English translation of the 'toxicity_cn' column.
- therapeutic_cn: The therapeutic category of the herb in Chinese, providing a cultural and linguistic context for its use.
- therapeutic_en: The therapeutic category of the herb in English, which may include terms like 'blood activation and stasis removal' or 'medicinal for detoxification'.
- HERB_ID: A cross-reference identifier linking the herb to the HERB database.
- TCMID_id: A cross-reference identifier linking the herb to the TCMID database.
- TCM-ID_id: A cross-reference identifier for the herb within the TCM-ID database.
- SymMap_id: A cross-reference identifier for the herb within the SymMap database.
- TCMSP_id: A cross-reference identifier for the herb within the TCMSP database.

### Ingredients - 1283

The ingredient info file includes all detail information of the 1,283 ingredients. The detail of each column are as follows:

- Ingredient_name: The common name by which the ingredient is widely recognized.
- Ingredient_alias_name: A list of alternative names or aliases for the ingredient, separated by semicolons.
- Molecular_formula: The chemical formula representing the molecule's composition.
- Canonical_smiles: A unique representation of the molecular structure in the SMILES (Simplified Molecular-Input Line-Entry System) format.
- Isomeric_smiles: A SMILES string that may differ from the canonical SMILES due to isomerism in the molecule's structure.
- InChIKey: A hashed version of the InChI, used as a compact and web-friendly identifier for the substance.
- MolWt: The molecular weight of the ingredient, a measure of the mass of one mole of the substance.
- NumHAcceptors: The number of hydrogen bond acceptors in the molecule.
- NumHDonors: The number of hydrogen bond donors in the molecule.
- MolLogP: The logarithm of the octanol-water partition coefficient, a measure of the molecule's hydrophobicity.
- NumRotatableBonds: The number of rotatable bonds in the molecule, which can affect its flexibility and pharmacokinetics.
- Drug_likeness: A quantitative measure estimating how closely the ingredient resembles drugs.
- OB_score: The Oral Bioavailability score, indicating the likelihood of the substance being absorbed in the gastrointestinal tract when taken orally.
- CAS_id: A cross-reference identifier for the ingredient in the Chemical Abstracts Service database.
- SymMap_id: A cross-reference identifier linking the ingredient to the SymMap database, which may provide information on the ingredient's effects on biological systems.
- TCMID_id: A cross-reference identifier for the ingredient in the TCMID database.
- TCMSP_id: A cross-reference identifier for the ingredient in the TCMSP database.
- TCM_ID_id: A cross-reference identifier for the ingredient in the TCM-ID database.
- HERB_id: A cross-reference identifier linking the ingredient to the HERB database.
- PubChem_id: A cross-reference identifier for the ingredient in the PubChem database.
- DrugBank_id: A cross-reference identifier for the ingredient in the DrugBank database.
- NPASS_id: A cross-reference identifier for the ingredient in the NPASS database.
- HIT_id: A cross-reference identifier for the ingredient in the HIT 2.0 database.

### References - 6705

The reference info file includes all detail information of the `6,705 papers` records related to drugs (herbs, ingredients, or formulae) in HERB database. The detail of each column are as follows:

- PubMed_id: The identifier for the reference as recorded in the PubMed database, a public database of biomedical literature.
- Subject_name: The name of the herb, ingredient, or formula that is the focus of the reference.
- Subject_type: The type of the subject, which can be an herb, ingredient, or formula.
- Paper_title: The title of the paper or article that the reference pertains to.
- Paper_abstract: A brief summary of the paper's content, including its aims, methods, and findings.
- Journal: The name of the journal in which the paper was published.
- DOI: A unique alphanumeric string assigned to the paper, which provides a persistent link to its location on the internet.
- Publish_date: The date on which the paper was published.
- Experiment_subject: The entity or model used in the experiments described in the paper, such as cell lines, animals, or human subjects.
- Experiment_type: The type of experiment conducted, which could include in vitro studies, animal experiments, clinical trials, etc.
- Phenotype_related: The disease or phenotype that is relevant to the research presented in the paper.

### Experiments - 2231

The experiment info file includes all detail information of the 2,231 experiments related to drugs (herbs, ingredients, or formulae) and diseases. The detail of each column are as follows:

- Subject_disease_name: The common name of the subject or the disease associated with the experiment.
- Subject_disease_id: The unique identifier for the subject or the disease in HERB database.
- Subject_type: The type of the subject, such as 'Herb', 'Ingredient', 'Disease', or 'Formula'.
- GSE_id: The identifier from the NCBI Gene Expression Omnibus (GEO) database for the experiment.
- Organism: The biological organism used in the experiment, for example, 'Homo sapiens' for humans.
- Experiment_type: The general type of experiment conducted, such as 'Expression profiling by array' or 'Expression profiling by high throughput sequencing'.
- Sequence_type: The type of sequencing used in the experiment, if applicable, such as 'total RNA'.
- Experiment_subject: The class of subjects involved in the experiment, which could be humans, animals, cell lines, or tissues.
- Experiment_detail: Additional details about the subjects involved in the experiment.
- Special_pretreatment: Any special pretreatments or conditions that the subjects were subjected to as part of the experiment.
- Control_condition: The conditions under which the control group was kept.
- Control_samples: The samples that serve as the control group in the experiment.
- Experiment_subject_detail: Further details about the subjects of the experiment, such as specific treatments or conditions.
- Treatment_samples: The samples that received the treatment or intervention in the experiment.
- Drug_delivery: The method by which drugs or other treatments were delivered to the subjects.
- Data_type: The data type of the experiment.
- Data_process_pipeline: The specific pipeline used for data processing.
- Platform: Refers to the specific GPL identifier used in the study to denote the microarray platform or sequencing technology utilized for the gene expression analysis.
- Original_id: The original identifier for the experiment in the source database or study.
- Exp_number: A set of experiment identifiers within the same GSE that correspond to a group of experiments under the same experimental conditions, allowing for the grouping and comparison of similar experimental setups.
- Strain: The specific strain of an animal model used in the experiment, if applicable.
- Tissue: The type of tissue sampled in the experiment, if applicable.
- Cell_type: The type of cells used in the experiment, if applicable.
- Cell_line: The specific cell line used in the experiment, if applicable.

## benchmark

In order to further effectively and accurately assess the performance of large models in the field of traditional Chinese medicine (TCM), we have now established a standardized and comprehensive TCM evaluation framework. This evaluation framework will fully take into account the complexity and professionalism of the TCM field, covering multiple aspects to ensure the practicality and applicability of large language models in real-world scenarios.

### Dataset

The dataset includes three categories of multiple-choice questions: Traditional Chinese Medicine (TCM) Diagnosis, TCM Pediatrics, and Internal Medicine. Each category contains up to 100 questions for each of the 16 knowledge points. We will gradually release more test questions and expand this dataset in the future. By collecting, organizing, and annotating these data, we aim to provide a comprehensive, accurate, and representative TCM testing benchmark to help evaluate and improve the performance of large language models in the field of TCM.

### Question Types

- **Best Choice Questions or Single-Choice Questions (Type A)**: Each question consists of a stem and five optional answers labeled A, B, C, D, and E. The stem comes first, followed by the five options. Only one of the five options is the correct answer. Type A questions are divided into two subtypes: Type A1 and Type A2.
- **Single-Sentence Best Choice Questions (Type A1)**: The stem is presented in the form of a statement, which can be affirmative or negative.
- **Case Summary Best Choice Questions (Type A2)**: The stem is a brief case description.
Case Group Best Choice Questions (Type A3): The stem begins with a clinical scenario centered on a patient, followed by several (more than one) related questions. Each question is connected to the initial clinical scenario but tests different key points. The focus of these questions is on clinical application.
- **Standard Matching Questions (Type B1)**: Each question consists of five optional answers labeled A, B, C, D, and E, followed by two or more stems. When answering, one option must be selected for each stem. Each optional answer can be used zero, one, or multiple times.

üí• Note: Since single-sentence best choice questions and case summary best choice questions have similar answer structures, they are combined into one type during testing and labeled as Single-Stem Single-Best-Choice Questions (Type A1/A2).

The statistics of benchmark are as follows:

| Question Type          | Single-Stem Single-Best-Choice Questions (Type A1/A2)| Case Group Best Choice Questions (Type A3)|Standard Matching Questions (Type B1)|
| ------------------ | -------------- | -------------- |-------------- |
| **Number of Questions**           | 1600           | 198         |1481         |
| **Sub-questions**             | \           | 642          | 3231          |

### Data Processing

The questions are transformed into structured evaluation data, with the format shown below:
Single-Stem Single-Best-Choice Questions (Type A1/A2):

```json
 {
      "question": "„ÄäÁ¥†ÈóÆ¬∑Âí≥ËÆ∫„ÄãÔºö‚Äú‰∫îËÑèÂÖ≠ËÖëÁöÜ‰ª§‰∫∫Âí≥‚ÄùÔºå‰ΩÜÂÖ≥Á≥ªÊúÄÂØÜÂàáÁöÑÊòØÔºà  Ôºâ„ÄÇ\nAÔºéÂøÉËÇ∫\nBÔºéËÇ∫ËÇæ\nCÔºéËÇ∫ËÑæ\nDÔºéËÇ∫ËÉÉ\nEÔºéËÇ∫Â§ßËÇ†",
      "answer": [
        "D"
      ],
      "analysis": "Ê†πÊçÆ„ÄäÁ¥†ÈóÆ¬∑Âí≥ËÆ∫„Äã‚ÄúÊ≠§ÁöÜËÅö‰∫éËÉÉÔºåÂÖ≥‰∫éËÇ∫Ôºå‰Ωø‰∫∫Â§öÊ∂ïÂîæËÄåÈù¢ÊµÆËÇøÊ∞îÈÄÜ‰πü‚ÄùÂèØÁü•‰∏é‰∫îËÑèÂÖ≠ËÖëÁöÜ‰ª§‰∫∫Âí≥ÂÖ≥Á≥ªÊúÄÂØÜÂàáÁöÑËÑèËÖë‰∏∫ËÇ∫ËÉÉ„ÄÇÊâãÂ§™Èò¥ËÇ∫ÁªèËµ∑‰∫é‰∏≠ÁÑ¶ÔºåËøòÂæ™ËÉÉÂè£Ôºå‰∏äËÜàÂ±ûËÇ∫„ÄÇÂØíÂáâÈ•ÆÈ£üÂÖ•ËÉÉÔºåÂØºËá¥‰∏≠ÁÑ¶ÂØíÔºåÂØíÊ∞îÂæ™ÊâãÂ§™Èò¥ËÇ∫Áªè‰∏äÂÖ•‰∫éËÇ∫‰∏≠ÔºåÂØºËá¥ËÇ∫ÂØíÔºåËÇ∫‰∏∫Â®áËÑèÔºå‰∏çËÄêÂØíÁÉ≠ÔºåÂ§ñÂÜÖÂØíÈÇ™Âπ∂ËÅö‰∫éËÇ∫ÔºåÂàôËÇ∫Â§±ÂÆ£ÈôçÔºåËÇ∫Ê∞î‰∏äÈÄÜÂèëÁîüÂí≥ÂóΩ„ÄÇÂõ†Ê≠§Á≠îÊ°àÈÄâD„ÄÇ",
      "knowledge_point": "‰∏≠ÂåªÁªèÂÖ∏",
      "index": 8196,
      "score": 1
    }
```

Case Group Best Choice Questions (Type A3):

```json
    {
      "share_content": "Âàò√óÔºåÁî∑Ôºå46Â≤ÅÔºåÂàª‰∏ãÁú©ÊôïËÄåËßÅÂ§¥ÈáçÂ¶ÇËíô„ÄÇËÉ∏Èó∑ÊÅ∂ÂøÉÔºåÈ£üÂ∞ëÂ§öÂØêÔºåËãîÁôΩËÖªÔºåËÑâÊø°Êªë„ÄÇ",
      "question": [
        {
          "sub_question": "1)ÔºéËØÅÂ±ûÔºà  Ôºâ„ÄÇ\nAÔºéËÇùÈò≥‰∏ä‰∫¢\nBÔºéÊ∞îË°Ä‰∫èËôö\nCÔºéËÇæÁ≤æ‰∏çË∂≥\nDÔºéÁó∞Êµä‰∏≠Èòª\nEÔºé‰ª•‰∏äÈÉΩ‰∏çÊòØ\n",
          "answer": [
            "D"
          ],
          "analysis": ""
        },
        {
          "sub_question": "2)ÔºéÊ≤ªÊ≥ïÂÆúÈÄâÔºà  Ôºâ„ÄÇ\nAÔºéÁá•ÊπøÁ•õÁó∞ÔºåÂÅ•ËÑæÂíåËÉÉ\nBÔºéË°•ËÇæÊªãÈò¥\nCÔºéË°•ËÇæÂä©Èò≥\nDÔºéË°•ÂÖªÊ∞îË°ÄÔºåÂÅ•ËøêËÑæËÉÉ\nEÔºéÂπ≥ËÇùÊΩúÈò≥ÔºåÊªãÂÖªËÇùËÇæ\n",
          "answer": [
            "A"
          ],
          "analysis": ""
        },
        {
          "sub_question": "3)ÔºéÊñπËçØÂÆúÈÄâÔºà  Ôºâ„ÄÇ\nAÔºéÂè≥ÂΩí‰∏∏\nBÔºéÂ∑¶ÂΩí‰∏∏\nCÔºéÂçäÂ§èÁôΩÊúØÂ§©È∫ªÊ±§\nDÔºéÂΩíËÑæÊ±§\nEÔºéÂ§©È∫ªÈí©Ëó§È•Æ\n",
          "answer": [
            "C"
          ],
          "analysis": ""
        }
      ],
      "knowledge_point": "‰∏≠ÂåªÂÜÖÁßëÂ≠¶",
      "index": 334,
      "score": 1
    }
```

Standard Matching Questions (Type B1):

```json
  {
      "share_content": "ÔºàÂÖ±Áî®Â§áÈÄâÁ≠îÊ°àÔºâ\nA.ÂåñÁó∞ÊÅØÈ£éÔºåÂÅ•ËÑæÁ•õÊπø\nB.Ê∏ÖËÇ∫ÂåñÁó∞ÔºåÊï£ÁªìÊéíËÑì\nC.ÁñèÈ£éÂÆ£ËÇ∫ÔºåÂåñÁó∞Ê≠¢Âí≥\nD.Ê∏ÖÁÉ≠ÂåñÁó∞ÔºåÂπ≥ËÇùÊÅØÈ£é\nE.Ê∂¶ËÇ∫Ê∏ÖÁÉ≠ÔºåÁêÜÊ∞îÂåñÁó∞\n",
      "question": [
        {
          "sub_question": "1)ÔºéË¥ùÊØçÁìúËíåÊï£ÁöÑÂäüÁî®ÊòØÔºà  Ôºâ„ÄÇ",
          "answer": [
            "E"
          ],
          "analysis": ""
        },
        {
          "sub_question": "2)ÔºéÂçäÂ§èÁôΩÊúØÂ§©È∫ªÊ±§ÁöÑÂäüÁî®ÊòØÔºà  Ôºâ„ÄÇ",
          "answer": [
            "A"
          ],
          "analysis": ""
        }
      ],
      "knowledge_point": "ÊñπÂâÇÂ≠¶",
      "index": 1938,
      "score": 1
    }
```

### Prompt Evaluation

We have designed different prompts for each question type, requiring the LLM to answer the questions and provide answers and analyses. The evaluation framework consists of the following components:

| File Name| Description|
| ---| --- |
| /evaluate/A12_bench.py     | Generates answers for single-stem single-best-choice questions |
| /evaluate/A3-B1_bench.py      | Generates answers for case-based best-choice questions and standard matching questions|
| /evaluate/bench_function.py   | Test-related functions |
| /evaluate/correct_analyse.py  | Calculates accuracy |
| /prompt/A1-2_prompt.json| Instruction file for single-stem single-best-choice questions (A1/A2 type)|
| /prompt/A3-4_prompt.json| Instruction file for case-based best-choice questions (A3 type)|
| /prompt/B1_prompt.json| Instruction file for standard matching questions (B1 type)| 
| /models/Openai.py| Calls the OpenAI API |

The fields in the instruction files are as follows:

| Field           | Description|
| --- | --- |
| **type** |Question type|
| **keyword** |Dataset search keyword|
| **prefix_prompt**| Instruction information|

You can generate answers for the three types of questions by running `A12_bench.py` and `A3-B1_bench.py` using different model APIs. We have provided the OpenAI API in the `/models` folder. Other custom model APIs can be placed in this folder for calling.

```python
# First, if necessary, set up a proxy:
os.environ['HTTPS_PROXY']="your proxy"
# Next, fill in your OpenAI Key in the designated location:
openai_api_key = "your key"
# Then, call different models by setting different model_type and model_name.
# The following commands have been run:
python A12_bench.py
python A3-B1_bench.py
```

Finally, you can run `correct_analyse.py` to obtain the model's accuracy score.

```
python correct_analyse.py
```

## Citation

If you use the data from this project, please cite it as follows:

```bash
@misc{cai2025 SOAP,
      title={SOAP:Semantic-Oriented Alignment for Prescription Generation}, 
      author={Jiale Cai},
      year={2025},
      publisher = {GitHub},
      journal = {GitHub repository},
      howpublished = {\url{https://github.com/sheldoncoder1337/SOAP}},
}
```
